File: C++/Dynamic Programming/10. Regular Expression Matching.cpp
#include <unordered_map>

struct PairHash{
    std::size_t operator()(const pair<int, int>& p) const {
        auto h1 = std::hash<int>{}(p.first);
        auto h2 = std::hash<int>{}(p.second);
        return h1^h2;
    }
};

class Solution {
    unordered_map<pair<int, int>, bool, PairHash> memo;
    string text;
    string pattern;
    
public:
    bool helper(int i, int j) {
        bool res;
        bool firstMatch;
        if (memo.find({i, j}) == memo.end()) {
            if (j == pattern.size()) {
                res = i == text.size();
            }
            else {
                firstMatch = i < text.size() && (pattern[j] == text[i] || pattern[j] == '.');
                if (j+1 < pattern.size() && pattern[j+1] == '*') {
                    res = helper(i, j+2) || (firstMatch && helper(i+1, j));
                }
                else {
                    res = firstMatch && helper(i+1, j+1);
                }
            }
            memo[{i, j}] = res;
        }
        return memo[{i, j}];
    }
    
    bool isMatch(string s, string p) {
        text = s;
        pattern = p;
        return helper(0, 0);
    }
};

File: Python/10. Regular Expression Matching.py
class Solution(object):
    def helper(self, i, j):
        if (i, j) not in self.memo:
            if j == len(self.pattern):
                ans = i == len(self.text)
            else:
                first_match = i < len(self.text) and self.pattern[j] in {self.text[i], '.'}
                if j + 1 < len(self.pattern) and self.pattern[j+1] == "*":
                    ans = self.helper(i, j+2) or first_match and self.helper(i+1, j)
                else:
                    ans = first_match and self.helper(i+1, j+1)
            self.memo[i, j] = ans
        return self.memo[i, j]
        
                
    def isMatch(self, text, pattern):
        self.memo = {}
        self.text = text
        self.pattern = pattern
        
        return self.helper(0, 0)